Implemented upload session APIs (multi-part uploads)

jpillora opened the pull request:
Changes
- implemented upload session APIs (multi-part uploads)
- ~~implemented auto-handled upload session method `UploadSession()`~~ (see go-dropy)
- renamed internal `download()` to `content()` ("Content-upload endpoints" and "Content-download endpoints")
- added internal `decodeContent()` helper
- fixed `PermanentlyDelete` endpoint path (not sure if it was incorrect on purpose?)

Notes
1. Methods should probably use value parameters, as pointer parameters allow property changes after the call has been made. Also, it's good practise to use values where possible. However, I thought it was more important to follow the current method layout.
2. As an alternate API to reduce the size of `files.go`, the raw upload session methods (start/append/finish) could be moved to `files_upload_session.go` and then would become `Files.UploadSession.Start()`, `.Append()`, `.Finish()` - though we'd have to rename `UploadSession()` to something else.
3. As an additional feature, we could add a `PerformRetries bool` so if any chunk failed, it would be retransmitted. This would require a `ChunkSize` buffer though to hold data in case of failure and maybe the Dropbox API doesn't fail that often so it might not be worth it...

jpillora committed:
added upload session APIs (multi-part uploads) with test, renamed download() to content()

jpillora commented:
```
# cd .; git clone https://github.com/davecgh/go-spew /home/runner/workspace/src/github.com/davecgh/go-spew
Cloning into '/home/runner/workspace/src/github.com/davecgh/go-spew'...
error: RPC failed; curl 56 GnuTLS recv error (-9): A TLS packet with unexpected length was received.
```

Strange failure, maybe manually retry?

jpillora commented:
Ah yes, just remembered https://github.com/tj/go-dropbox/issues/2. So maybe the higher-level `UploadSession()` should go in `go-dropy`?

tj commented:
niceee, yea as much as possible I think the fancier stuff should be in go-dropy

jpillora committed:
strip out high-level UploadSession() method

moved to go-dropy, see https://github.com/jpillora/go-dropy/commit/5eee26d6380c7f4bd609f1cbdeb95e03355e2f24

jpillora committed:
fix Upload()

jpillora commented:
Striped out high-level `UploadSession()` method, moved to go-dropy, see jpillora/go-dropy@5eee26d - Will PR it over there once this lands.

jpillora commented:
Adding new test with manually managed session upload...

jpillora committed:
added upload start/append/finish (manual session) test

jpillora commented:
Semaphore seems to be a bit buggy, https://semaphoreci.com/tj/go-dropbox/branches/pull-request-18/builds/4 also failed strangely

jpillora committed:
sync with upstream

jpillora committed:
make semaphore badge match godoc badge

jpillora commented:
I'll sync with upstream again once #17 tests are fixed
